### PR TITLE
Postprocess aerial images

### PR DESCRIPTION
**Pull request by delhomer**
This PR introduces the postprocessing steps associated to semantic segmentation predictions over aerial georeferenced images.

The label inference is done on small image tiles, that must be merged once predicted labels are ready for every tile. As input, we get tiled images and the corresponding predictions. As output, we have a full labelled image directly comparable with original high-resolution image.

In more details a new module that undertake this operation is provided. It is inspired from existing `inference.py` module, however it focuses only on aerial image datasets (`AerialImage` and `Tanzania`, for now) and semantic segmentation models.

Additionally, a full unit test suite is provided along with the new module.

As complementary remarks, one may note that this PR introduces some duplicated code (`get_image_features(.)` appears twice, though it has two different signatures; then some code lines shall be merged between `inference.py` and `postprocess.py`, it may rely on issue #90). I propose to address it in a further PR.

### CODE REVIEW

File: deeposlandia/postprocess.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+resolution image.
+
+Do inference on test images starting from a trained model and Keras API. Then
+handle geographic datasets with dedicated libraries (GDAL, geopandas, shapely)
+"""
+
+import argparse
+import glob
+import os
+import sys
```
Comment (garaud) [inline deeposlandia/postprocess.py]: `sys` is imported but not used

File: deeposlandia/postprocess.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    prepro_folder = utils.prepare_preprocessed_folder(
+        datapath, dataset, tile_size, "full"
+    )
+    if os.path.isfile(prepro_folder["testing_config"]):
+        test_config = utils.read_config(prepro_folder["testing_config"])
+    else:
+        raise ValueError(("There is no testing data with the given "
+                          "parameters. Please generate a valid dataset "
+                          "before calling the program."))
+    labels = [l for l in test_config["labels"] if l["is_evaluate"]]
```
Comment (garaud) [inline deeposlandia/postprocess.py]: no need to create a variable. I think you can return the list directly

File: deeposlandia/postprocess.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+        List of dictionnaries that describes the dataset labels
+
+    Returns
+    -------
+    numpy.array
+        Coloured predicted labels, of shape (img_size, img_size, 3)
+    """
+    labelled_images = np.zeros(shape=np.append(predicted_labels.shape, 3),
+                               dtype=np.uint8)
+    for i in range(len(labels)):
```
Comment (garaud) [inline deeposlandia/postprocess.py]: I like this form

```python
for idx, label in enumerate(labels):
    labelled_images[predicted_labels == idx] = label["color"]
```

but it's not very important.

File: deeposlandia/postprocess.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+
+    Returns
+    -------
+    list
+        List of tiled image west and north pixel coordinates, regarding the
+    full original image
+    """
+    basenames = [os.path.splitext(os.path.basename(f))[0] for f in filenames]
+    coordinates = [b.split("_")[-2:] for b in basenames]
+    return [[int(c[0]), int(c[1])] for c in coordinates]
```
Comment (garaud) [inline deeposlandia/postprocess.py]: maybe you can write `[[int(x), int(y)] for x, y in coordinates]`

File: deeposlandia/postprocess.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+    ds = gdal.Open(filepath)
+    width = ds.RasterXSize
+    height = ds.RasterYSize
+    gt = ds.GetGeoTransform()
+    minx = gt[0]
+    miny = gt[3] + height * gt[5]
+    maxx = gt[0] + width * gt[1]
+    maxy = gt[3]
+    srid = int(ds.GetProjection().split('"')[-2])
+    ds = None
```
Comment (garaud) [inline deeposlandia/postprocess.py]: maybe you can leave a comment about "why do you set None to ds"

File: tests/test_postprocess.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        postprocess.get_labels(
+            "./elsewhere-on-the-disk", "wrong-dataset", tanzania_image_size
+        )
+
+
+def test_get_trained_model(tanzania_image_size, tanzania_nb_labels):
+    """Test the semantic segmentation model retrieving
+
+    Postprocessing implies getting a 'unet' model (i.e. segmentation
+    segmentation), hence the model must be shaped as follows:
```
Comment (garaud) [inline tests/test_postprocess.py]: typo. you wrote "segmentation segmentation" :slightly_smiling_face:

File: tests/test_postprocess.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    model_conv_layers = [ml for ml in model.layers if "conv" in ml.name]
+    assert len(model_conv_layers) == 57
+    model_up_layers = [ml for ml in model.layers if "up" in ml.name]
+    assert len(model_up_layers) == 8
+    model_pool_layers = [ml for ml in model.layers if "pool" in ml.name]
+    assert len(model_pool_layers) == 4
+
+
+def test_assign_label_colors():
+    """Test the label colourization function, that allows to replace label IDs
```
Comment (garaud) [inline tests/test_postprocess.py]: colorization?

File: tests/test_postprocess.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    with pixel triplets
+
```
Comment (garaud) [inline tests/test_postprocess.py]: remaining empty lines

File: tests/test_postprocess.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    """
+    labels = [{"name": "foo", "color": [0, 0, 0]},
+              {"name": "bar", "color": [200, 200, 200]}]
+    y = np.array([
+        [[1, 1, 0], [1, 1, 0], [0, 1, 1]],
+        [[0, 1, 1], [0, 0, 0], [0, 1, 1]]
+    ])
+    l1 = np.array([
```
Comment (garaud) [inline tests/test_postprocess.py]: sometimes, I use the variable name `expected` to compare the result of the function with the expected values

File: tests/test_postprocess.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+                   [x3, y3], [x4, y4]]
+    e1 = np.array(
+        [[1, 2, 5, 6],
+         [3, 4, 7, 8],
+         [9, 10, 13, 14],
+         [11, 12, 15, 16]],
+        dtype=np.int8
+    )
+    e2 = postprocess.fill_labelled_image(tiles, coordinates, 2, 4)
+    print(e1)
```
Comment (garaud) [inline tests/test_postprocess.py]: remaining prints

File: tests/test_postprocess.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+    e1 = np.array(
+        [[1, 2, 5],
+         [3, 4, 7],
+         [9, 10, 13],
+         [11, 12, 15],
+         [17, 18, 21]],
+        dtype=np.int8
+    )
+    e2 = postprocess.fill_labelled_image(tiles, coordinates, 2, 3, 5)
+    print(e1)
```
Comment (garaud) [inline tests/test_postprocess.py]: remaining prints

**Review by garaud — CHANGES_REQUESTED** [review]

File: deeposlandia/postprocess.py (hunk 1 above)
Comment (delhomer) [reply deeposlandia/postprocess.py]: The useless import has been removed.

File: deeposlandia/postprocess.py (hunk 2 above)
Comment (delhomer) [reply deeposlandia/postprocess.py]: Simplified!

File: deeposlandia/postprocess.py (hunk 3 above)
Comment (delhomer) [reply deeposlandia/postprocess.py]: That's more elegant, no problem for changing this.

File: deeposlandia/postprocess.py (hunk 4 above)
Comment (delhomer) [reply deeposlandia/postprocess.py]: Modified.

File: deeposlandia/postprocess.py (hunk 5 above)
Comment (delhomer) [reply deeposlandia/postprocess.py]: Added in the docstring: "A GDAL dataset is opened during the function execution. The corresponding variable is set to None at the end of the function so as to free memory."
Added as a comment at this line: "# Free memory used by the GDAL Dataset"

File: tests/test_postprocess.py (hunk 7 above)
Comment (delhomer) [reply tests/test_postprocess.py]: You're right. I write the code and the docstrings in US english, however "colourization" comes from UK english. I change that!

File: tests/test_postprocess.py (hunk 8 above)
Comment (delhomer) [reply tests/test_postprocess.py]: Removed.

File: tests/test_postprocess.py (hunk 6 above)
Comment (delhomer) [reply tests/test_postprocess.py]: A new AI model, you don't know it?
...
Modified.

File: tests/test_postprocess.py (hunk 9 above)
Comment (delhomer) [reply tests/test_postprocess.py]: Good remark. By the way, the last unit tests I wrote contain "expected_something" variables... I change this occurrence.

File: tests/test_postprocess.py (hunk 10 above)
Comment (delhomer) [reply tests/test_postprocess.py]: Removed.

File: tests/test_postprocess.py (hunk 11 above)
Comment (delhomer) [reply tests/test_postprocess.py]: Removed.

**Review by garaud — APPROVED** [review]

**Pull request merged**